internal/secrets/secrets.go

```diff
@@ -0,0 +1,219 @@
+package secrets
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/alecthomas/kingpin/v2"
+)
+
+type keyType int
+
+// Secret types.
+const (
+	Unset keyType = iota
+	Command
+	Config
+	EnvVar
+	File
+	Keychain
+	Value
+	Vault
+)
+
+// Secret holds secrets.
+type Secret struct {
+	Input      string
+	Value      string
+	StoreValue string
+	Type       keyType
+}
+
+// NewSecret constructs a new Secret.
+func NewSecret(value string) *Secret {
+	s := Secret{}
+
+	err := s.Set(value)
+	if err != nil {
+		return nil
+	}
+
+	return &s
+}
+
+// Set will set the Secret type.
+func (s *Secret) Set(value string) error {
```

Comment:
can we follow the pattern we have elsewhere (encryption, hash, compression, blob storage, etc.) where there's package-wide registry of supported algorithms and a way to register one on startup

The use case here is to allow folks extending Kopia to register completely new secret storage mechanisms that should just work.

Reply:
what I meant was exposing something like `secrets.Register(prefix, impl)` so you can plug in various implementations:

```
secrets.Register("command", &commandProvider{})
secrets.Register("envvar", &envvarProvider{})
secrets.Register("vault", &vaultProvider{})
secrets.Register("keyring", &keyringProvider{})
```